File: src/research_artifact_uploader/oss.py
from __future__ import annotations

import re
import shutil
from dataclasses import dataclass
from pathlib import Path

from .subprocess_utils import run_capture, run_stream


@dataclass
class OssUploadResult:
    success: bool
    avg_mib_s: float | None
    stdout: str
    stderr: str
    command: list[str]


def _require_ossutil() -> None:
    if not shutil.which("ossutil"):
        raise FileNotFoundError("ossutil not found")


def _parse_avg_speed(text: str) -> float | None:
    patterns = [
        r"avg\s+([0-9]+(?:\.[0-9]+)?)\s*MiB/s",
        r"average\s+speed:\s*([0-9]+(?:\.[0-9]+)?)\s*MiB/s",
        r"avg\s+([0-9]+(?:\.[0-9]+)?)\s*MB/s",
    ]
    for p in patterns:
        m = re.search(p, text, flags=re.IGNORECASE)
        if m:
            return float(m.group(1))
    return None


def oss_uri(bucket: str, remote_dir: str, filename: str) -> str:
    return f"oss://{bucket}/{remote_dir.rstrip('/')}/{filename}"


def check_access(bucket: str, endpoint: str, region: str) -> tuple[bool, str]:
    _require_ossutil()
    cmd = ["ossutil", "ls", f"oss://{bucket}", "-e", endpoint, "--region", region]
    result = run_capture(cmd)
    if result.returncode == 0:
        return True, result.stdout
    return False, (result.stdout or result.stderr or "").strip()


def upload_file(
    local_path: Path,
    bucket: str,
    remote_dir: str,
    filename: str,
    endpoint: str,
    region: str,
) -> OssUploadResult:
    _require_ossutil()
    uri = oss_uri(bucket, remote_dir, filename)
    cmd = [
        "ossutil",
        "cp",
        str(local_path),
        uri,
        "-e",
        endpoint,
        "--region",
        region,
    ]
    result = run_stream(cmd, stream=True)
    if result.returncode != 0:
        command_str = " ".join(cmd)
        stdout_snip = (result.stdout or "").replace("\\n", "\\n")[:800]
        raise RuntimeError(
            "ossutil cp failed (exit code {}).\ncommand: {}\nerror: {}\nsuggestion: retry command manually after fixing the cause.".format(
                result.returncode,
                command_str,
                stdout_snip or "(no output)",
            )
        )
    return OssUploadResult(
        success=True,
        avg_mib_s=_parse_avg_speed(result.stdout),
        stdout=result.stdout,
        stderr=result.stderr,
        command=result.command,
    )

File: src/oss.rs
use std::{path::Path, sync::OnceLock};
use regex::Regex;

use crate::{AppResult, manifest::Manifest, subprocess_utils::{run_capture, run_stream}};

#[derive(Debug)]
pub struct OssUploadResult {
    pub success: bool,
    pub avg_mib_s: Option<f64>,
    pub stdout: String,
    pub stderr: String,
    pub command: Vec<String>,
}

static RE_MIB: OnceLock<Regex> = OnceLock::new();

pub fn oss_uri(bucket: &str, remote_dir: &str, filename: &str) -> String {
    format!("oss://{}/{}/{}", bucket, remote_dir.trim_start_matches('/'), filename)
}

pub fn check_access(bucket: &str, endpoint: &str, region: &str) -> AppResult<(bool, String)> {
    let cmd = vec![
        "ossutil".to_string(),
        "ls".to_string(),
        format!("oss://{bucket}"),
        "-e".to_string(),
        endpoint.to_string(),
        "--region".to_string(),
        region.to_string(),
    ];

    let result = run_capture(&cmd, None).map_err(|e| format!("ossutil check failed: {e}"))?;
    if result.returncode == 0 {
        Ok((true, String::new()))
    } else {
        Ok((false, format!("{}", result.stdout.trim())))
    }
}

pub fn upload_file(
    local_path: &Path,
    manifest: &Manifest,
    remote_dir: &str,
    filename: &str,
) -> AppResult<OssUploadResult> {
    let uri = oss_uri(&manifest.oss.bucket, remote_dir, filename);
    let cmd = vec![
        "ossutil".to_string(),
        "cp".to_string(),
        local_path
            .to_str()
            .ok_or_else(|| "invalid local path".to_string())?
            .to_string(),
        uri,
        "-e".to_string(),
        manifest.oss.endpoint.clone(),
        "--region".to_string(),
        manifest.oss.region.clone(),
    ];

    let result = run_stream(&cmd, None).map_err(|e| format!("ossutil execution failed: {e}"))?;
    if result.returncode != 0 {
        let summary = (result.stderr.as_str()).trim();
        let summary = if summary.is_empty() {
            result.stdout.trim()
        } else {
            summary
        };
        let command = cmd.join(" ");
        return Err(format!(
            "ossutil cp failed (exit code {}).\ncommand: {command}\nerror: {}\nsuggestion: retry command manually after fixing the root cause.",
            result.returncode,
            summary,
        ));
    }

    Ok(OssUploadResult {
        success: true,
        avg_mib_s: parse_avg_speed(&result.stdout),
        stdout: result.stdout,
        stderr: result.stderr,
        command: result.command,
    })
}

fn parse_avg_speed(text: &str) -> Option<f64> {
    let re = RE_MIB.get_or_init(|| {
        Regex::new(r"(?i)(?:avg|average)\s+([0-9]+(?:\.[0-9]+)?)\s*MiB/s").unwrap()
    });
    re.captures(text)
        .and_then(|m| m.get(1))
        .and_then(|v| v.as_str().parse::<f64>().ok())
}
